File: Components/DevExpressVCL/Demos/ExpressRichEdit Control/CBuilder/Common/DBUriStreamProvider.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DBUriStreamProvider.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

_di_IdxUriStreamService GetUriStreamService(TdxRichEditControl* ARichEdit)
{
  TdxCustomDocumentModel* AServiceProvider = ARichEdit->InnerControl->DocumentModel;
  _di_IdxServiceProvider AServiceProviderIntf;
  AServiceProvider->GetInterface(AServiceProviderIntf);
  TGUID AGUID = Comobj::StringToGUID("{079683F3-DB6C-4DE9-A0AE-8EBB45900199}");
  return (_di_IdxUriStreamService)AServiceProviderIntf->GetService(AGUID);
}

_di_IdxUriStreamProvider GetUriStreamProvider(TClientDataSet* ADataSet,
	UnicodeString AKey, UnicodeString AField, UnicodeString APrefix)
{
  TdxDBUriStreamProvider* AUriProvider = new TdxDBUriStreamProvider(ADataSet, AKey, AField, APrefix);
  _di_IdxUriStreamProvider AUriProviderIntf;
  AUriProvider->GetInterface(AUriProviderIntf);
  return AUriProviderIntf;
}

void RegistrationDBUriStreamProvider(TdxRichEditControl* ARichEdit, TClientDataSet* ADataSet,
	UnicodeString AKey, UnicodeString AField, UnicodeString APrefix)
{
  _di_IdxUriStreamService AService = GetUriStreamService(ARichEdit);
  _di_IdxUriStreamProvider AUriProviderIntf = GetUriStreamProvider(ADataSet, AKey, AField, APrefix);
  AService->RegisterProvider(AUriProviderIntf);
}

__fastcall TdxDBUriStreamProvider::TdxDBUriStreamProvider(TClientDataSet* ADataSet,
	UnicodeString AKey, UnicodeString AField, UnicodeString APrefix) : TCppInterfacedObject<IdxUriStreamProvider>()
{
  FDataSet = new TClientDataSet(NULL);
  LoadData(ADataSet);
  FField = AField;
  FKey = AKey;
  FPrefix = APrefix;
}

__fastcall TdxDBUriStreamProvider::~TdxDBUriStreamProvider()
{
  delete FDataSet;
}

void TdxDBUriStreamProvider::GetCloneData(TClientDataSet* ADataSet)
{
  ADataSet->CloneCursor(FDataSet, True);
}

void TdxDBUriStreamProvider::LoadData(TClientDataSet* ADataSet)
{
  FDataSet->CloneCursor(ADataSet, True);
}

TStream* __fastcall TdxDBUriStreamProvider::GetStream(const UnicodeString AUrl)
{
  TStream* Result = NULL;
  if (Pos(FPrefix, Trim(AUrl)) == 1)
  {
	int Id;
	TLocateOptions AOptions;
	UnicodeString AString = AUrl.Trim();
	AString = AString.SubString(9, AUrl.Length());
	if (!TryStrToInt(AString, Id))
	  return Result;
	TClientDataSet* ADataSet = new TClientDataSet(NULL);
	GetCloneData(ADataSet);
	if (ADataSet->Locate(FKey, Id, AOptions))
	{
	  TBlobField* AField = (TBlobField*)(ADataSet->FieldByName(FField));
	  Result = new TMemoryStream();
	  AField->SaveToStream(Result);
	  Result->Position = 0;
	}
	delete ADataSet;
  }
  return Result;
}

bool __fastcall TdxDBUriStreamProvider::IsAsyncLoadingSupported(const UnicodeString AUrl)
{
  return false;
}

File: Components/DevExpressVCL/Demos/ExpressRichEdit Control/Delphi/Common/DBUriStreamProvider.pas
unit DBUriStreamProvider;

{$I cxVer.inc}

interface

uses
  Classes, DBClient, dxRichEdit.Utils.UriStreamService, MidasLib;

{$I cxVer.Inc}

type
{ TdxDBUriStreamProvider }

  TdxDBUriStreamProvider = class(TInterfacedObject, IdxUriStreamProvider)
  private
    FDataSet: TClientDataSet;
    FField: string;
    FKey: string;
    FPrefix: string;
    procedure LoadData(ADataset: TClientDataSet);
  protected
    function GetStream(const AUri: string): TStream;
    function IsAsyncLoadingSupported(const AUri: string): Boolean;
  public
    constructor Create(ADataSet: TClientDataSet; AKey, AField, APrefix: string);
    destructor Destroy; override;
  end;

implementation

uses
  SysUtils, DB;

{ TdxDBUriStreamProvider }

constructor TdxDBUriStreamProvider.Create(ADataSet: TClientDataSet; AKey, AField, APrefix: string);
begin
  inherited Create;
  FDataSet := TClientDataSet.Create(nil);
  LoadData(ADataSet);
  FField := AField;
  FKey := AKey;
  FPrefix := APrefix;
end;

destructor TdxDBUriStreamProvider.Destroy;
begin
  FreeAndNil(FDataSet);
  inherited Destroy;
end;

procedure TdxDBUriStreamProvider.LoadData(ADataset: TClientDataSet);
begin
  FDataSet.CloneCursor(ADataset, True);
end;

function TdxDBUriStreamProvider.GetStream(const AUri: string): TStream;
var
  Id: Integer;
  AField: TBlobField;
begin
  TMonitor.Enter(FDataSet);
  try
    Result := nil;
    if Pos(FPrefix, Trim(AUri)) = 1 then
    begin
      if not TryStrToInt(Copy(Trim(AUri), 9, Length(AUri)), Id) then
        Exit;
      if FDataSet.Locate(FKey, Id, []) then
      begin
        AField := FDataSet.FieldByName(FField) as TBlobField;
        Result := TMemoryStream.Create;
        AField.SaveToStream(Result);
        Result.Position := 0;
      end;
    end;
  finally
    TMonitor.Exit(FDataSet);
  end;
end;

function TdxDBUriStreamProvider.IsAsyncLoadingSupported(const AUri: string): Boolean;
begin
  Result := False;
end;

end.
